test/test_type_promotion.py

```diff
@@ -138,39 +139,39 @@ def test_mixed_type_backward(self):
         f_dtypes = [torch.float, torch.double]
         f_dtypes = f_dtypes if self.device == 'cpu' else f_dtypes + [torch.half]
         i_dtypes = [torch.int, torch.long]
-        for f in ['add', 'sub', 'rsub', 'mul', 'div']:
+        for func in [torch.add, torch.sub, torch.rsub, torch.mul, torch.div]:
             for dtype1 in f_dtypes:
                 for dtype2 in (f_dtypes + i_dtypes):
                     x = torch.ones(10, requires_grad=True, dtype=dtype1, device=self.device)
                     y = torch.ones(10, dtype=dtype2, device=self.device)
-
-                    func = getattr(torch, f)
                     func(x, y).sum().backward()
 
-    # verifies that a.add(b) is the same as a.to(b.dtype).add(b) in cases
-    # where that should hold.
+    # verifies that torch.<op>(a, b) is the same as 
+    # torch.<op>(a.to(common_dtype), b.to(common_dtype)) in cases where that should hold.
     def test_many_promotions(self):
-        from_to = {
-            torch.float16: torch.float32,
-            torch.half: torch.float16,
-            torch.int: torch.long,
-            torch.uint8: torch.long,
-            torch.uint8: torch.float,
-            torch.int: torch.float,
-            torch.int: torch.double,
-            torch.int16: torch.long,
-            torch.float16: torch.double,
-            torch.bool: torch.long,
-            torch.bool: torch.float
-        }
-
-        for k, v in from_to.items():
-            a = torch.rand([3, 3], device=self.device).to(k)  # no _th_uniform for half on cpu.
-            b = torch.rand([3, 3], device=self.device).to(v)
-            c = a.add(b)
-            d = a.to(v).add(b)
-            self.assertEqual(c.dtype, d.dtype, message='from {} to {}'.format(k, v))
-            self.assertEqual(c, d, message='from {} to {}'.format(k, v))
+        all_dtypes = torch.testing.get_all_math_dtypes(self.device)
+        for dt1, dt2 in itertools.product(all_dtypes, all_dtypes):
+            for op in [torch.add, torch.sub, torch.mul, torch.div]:
+                for contiguous in [True, False]:
+                    common_dtype = torch.promote_types(dt1, dt2)
+                    if op == torch.sub and common_dtype != torch.bool:
+                        # Subtraction, the `-` operator, with a bool tensor is not supported.
+                        continue
+
+                    first = torch.rand([20, 20], device=self.device).to(dt1)  # no _th_uniform for half on cpu.
+                    second = torch.rand([20, 20], device=self.device).to(dt2)
+                    # test ops with non-contiguous tensors
+                    if not contiguous:
+                        first = first.t()
+                        second = second.t()
+                    self.assertEqual(first.is_contiguous(), contiguous)
+                    self.assertEqual(second.is_contiguous(), contiguous)
+                    # ensure no div-by-zero
+                    second[second == 0] = 5
```

Comment:
This is three operations; it might be nicer to just construct second as
```
second = torch.rand([20, 20], device=self.device).to(dt2) + 0.1
```
to avoid the divide-by-zero.

Reply:
hm.... a `+` would cause a promotion so we'd have the wrong dtype. There might be a way to do it but I think trying to make that work could end up being less clear to read

Reply:
Oh, that's a good point... how about using torch.clamp(second, min=0.1)?

Reply:
If that doesn't work then what you have now is fine

Reply:
rand/randn don't support dtype=int anyway (and min=0.1 would probably get truncated to 0 if it did). rearranged a bit though because I want to test a better range of values. I noticed rand().to(int) was giving all 0's and 1's and to(bool) was giving all true.